### PR TITLE
storage: addition of hierarchical directory

### LINKED ISSUE #64: storage: hierarchical directory structure

### PR DESCRIPTION
**Pull request by JavierDelgadoFernandez**
* Adds the possiblity to store the files grouped by file name in
  different directories. (closes #64)

Signed-off-by: Javier Delgado <javier.delgado.fernandez@cern.ch>

### CODE REVIEW

File: invenio_files_rest/storage.py [hunk 1]
```diff
@@ -158,10 +160,18 @@ def __init__(self, fileinstance, base_uri=None, filename='data'):
         self.filename = filename
         super(PyFilesystemStorage, self).__init__(fileinstance)
 
-    def make_path(self):
+    def make_path(self, path_dimensions=FILES_REST_STORAGE_PATH_DIMENSIONS,
```
Comment (jirikuncar) [inline invenio_files_rest/storage.py]: please use following pattern it respects the application config:

```python
def foo(bar=None):
    bar = bar or current_app.config['BAR_DEFAULT']
```

File: tests/test_storage.py [hunk 2]
```diff
@@ -64,10 +64,6 @@ def test_file_save(data, **kwargs):
         assert "md5:{0}".format(m.hexdigest()) == checksum
 
         assert size == len(data)
-        assert loc == join(
```
Comment (jirikuncar) [inline tests/test_storage.py]: why did you remove this test?

File: tests/test_tasks.py [hunk 3]
```diff
@@ -58,7 +58,6 @@ def test_migrate_file(app, db, dummy_location, extra_location, bucket,
 
     # Test pre-condition
     old_uri = obj.file.uri
-    assert join(bucket.location.uri, str(obj.file.id), 'data') == old_uri
```
Comment (jirikuncar) [inline tests/test_tasks.py]: ! https://github.com/inveniosoftware/invenio-files-rest/pull/68/files#r65142159

File: tests/test_storage.py (hunk 2 above)
Comment (JavierDelgadoFernandez) [inline tests/test_storage.py]: Now there is a dedicated test checking if the paths are generated correctly. I think it is not good to keep this checking here because, moreover, it should be adapted to use the new path creations, and that requires the develop of the same method which generates the paths here again... If you hardcode the expected path for testing here, it will crash if the user modify the configuration to something else.

Comment (lnielsen) [inline tests/test_storage.py]:
```python
join(dummy_location.uri, str(obj.file.id)[0:2], str(obj.file.id)[:2], "data")
```

File: tests/test_storage.py [hunk 4]
```diff
@@ -161,3 +157,31 @@ def test_storage_interface():
     pytest.raises(NotImplementedError, s.send_file)
     pytest.raises(NotImplementedError, s.save, None)
     pytest.raises(NotImplementedError, s.compute_checksum, None)
+
+
+def test_pyfilesystemstorage_make_path():
+    """Test path for files."""
+    f = PyFilesystemStorage(
+        FileInstance(uri="Test"), base_uri='N')
```
Comment (lnielsen) [inline tests/test_storage.py]: ``python
f = FileInstance.create(...)
f.id = UUID('deadbeef-....')
```

**Comment by JavierDelgadoFernandez** [comment]
@lnielsen Updated.
